src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1966,23 +1966,31 @@ case class SymRef(posn: Position, symbol: String) extends AST(posn) {
 case class If(pos: Position, cond: AST, thenTree: AST, elseTree: AST)
   extends AST(pos, Array(cond, thenTree, elseTree)) {
   override def typecheckThis(ec: EvalContext): BaseType = {
-    if (thenTree.`type` != elseTree.`type`)
-      parseError(s"expected same-type `then' and `else' clause, got `${ thenTree.`type` }' and `${ elseTree.`type` }'")
-    else
-      thenTree.`type`
+    (thenTree.`type`, elseTree.`type`) match {
+      case (thenType, elseType) if thenType == elseType => thenType
+      case (thenType: TNumeric, elseType: TNumeric)     => TNumeric.promoteNumeric(Set(thenType, elseType))
+      case _ =>
+        parseError(s"expected same-type `then' and `else' clause, got `${ thenTree.`type` }' and `${ elseTree.`type` }'")
+    }
   }
 
   def eval(ec: EvalContext): () => Any = {
     val f1 = cond.eval(ec)
     val f2 = thenTree.eval(ec)
     val f3 = elseTree.eval(ec)
+
+    // this is necessary to avoid serializing `this` which is not serializable
+    // (and it would be an unnecessary leak anyway)
+    val ifType = `type`
+
     () => {
+      def coerce(value: Any): Any = ifType match {
```

Comment:
`ifType` is invariant of the thunk, so you can lift the match out.  Something like:

```
  val coerce: (Any) => Any = `type`  match { ... }
  () => {
    val c = f()
    ...
```

Otherwise, looks great!